core/sessions/ldapauth/ldap.go

```diff
@@ -0,0 +1,640 @@
+	}
+	defer conn.Close()
+	// Attempt to LDAP Bind with user provided credentials
+	escapedEmail := ldap.EscapeFilter(strings.ToLower(email))
+	searchBaseDN := fmt.Sprintf("%s=%s,%s,%s", l.config.BaseUserAttr(), escapedEmail, l.config.UsersDn(), l.config.BaseDn())
+	if err := conn.Bind(searchBaseDN, password); err != nil {
+		l.lggr.Infof("Error binding user authentication request in TestPassword call LDAP Bind: %v", err)
+		return errors.New("Invalid credentials")
+	}
+	return nil
+}
+
+// CreateAndSetAuthToken generates a new credential token with the user role
+func (l *ldapAuthenticator) CreateAndSetAuthToken(user *sessions.User) (*auth.Token, error) {
+	newToken := auth.NewToken()
+
+	err := l.SetAuthToken(user, newToken)
+	if err != nil {
+		return nil, err
+	}
+
+	return newToken, nil
+}
+
+// SetAuthToken updates the user to use the given Authentication Token.
+func (l *ldapAuthenticator) SetAuthToken(user *sessions.User, token *auth.Token) error {
+	salt := utils.NewSecret(utils.DefaultSecretSize)
+	hashedSecret, err := auth.HashedSecret(token, salt)
+	if err != nil {
+		return errors.Wrap(err, "LDAPAuth SetAuthToken hashed secret error")
+	}
+
+	err = l.q.Transaction(func(tx pg.Queryer) error {
+		// Remove any existing API tokens
+		if _, err := l.q.Exec("DELETE FROM ldap_user_api_tokens WHERE user_email = $1"); err != nil {
+			return errors.Wrap(err, "Error executing DELETE FROM ldap_user_api_tokens")
+		}
+		// Create new API token for user
+		_, err = l.q.Exec(
+			"INSERT INTO ldap_user_api_tokens (user_email, user_role, token_key, token_salt, token_hashed_secret, created_at) VALUES ($1, $2, $3, $4, $5, now())",
+			user.Email,
+			user.Role,
+			token.AccessKey,
+			salt,
+			hashedSecret,
+		)
+		return nil
+	})
+	if err != nil {
+		return errors.Wrap(err, "error removing potential existing token and new token creation")
+	}
+
+	l.auditLogger.Audit(audit.APITokenCreated, map[string]interface{}{"user": user.Email})
+	return nil
+}
+
+// DeleteAuthToken clears and disables the users Authentication Token.
+func (l *ldapAuthenticator) DeleteAuthToken(user *sessions.User) error {
+	_, err := l.q.Exec("DELETE FROM ldap_user_api_tokens WHERE email = $1")
+	return err
+}
+
+// SaveWebAuthn is not supported for read only LDAP
+func (l *ldapAuthenticator) SaveWebAuthn(token *sessions.WebAuthn) error {
+	return ErrNotSupported
+}
+
+// Sessions returns all sessions limited by the parameters.
+func (l *ldapAuthenticator) Sessions(offset, limit int) ([]sessions.Session, error) {
+	var sessions []sessions.Session
+	sql := `SELECT * FROM ldap_sessions ORDER BY created_at, id LIMIT $1 OFFSET $2;`
+	if err := l.q.Select(&sessions, sql, limit, offset); err != nil {
+		return sessions, nil
+	}
+	return sessions, nil
+}
+
+// FindExternalInitiator supports the 'Run' role external intiator header auth functionality
+func (l *ldapAuthenticator) FindExternalInitiator(eia *auth.Token) (*bridges.ExternalInitiator, error) {
```

Comment:
Duplication felt better here since these modules are siblings in the hierarchy, importing from the other local ORM felt dirty. We could factor this out to the higher abstraction of types though